cns/fsnotify/fsnotify.go

```diff
@@ -0,0 +1,147 @@
+package fsnotify
+
+import (
+	"context"
+	"io/fs"
+	"os"
+
+	"github.com/Azure/azure-container-networking/cns"
+	"github.com/fsnotify/fsnotify"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+)
+
+type cnsclient interface {
+	ReleaseIPs(ctx context.Context, ipconfig cns.IPConfigsRequest) error
+}
+
+type Watcher struct {
+	CnsClient cnsclient
+}
+
+func WatchFs(w *Watcher, path string, logger *zap.Logger) error {
+	// Create new fs watcher.
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		logger.Error("error creating watcher", zap.Error(err))
+	}
+	defer watcher.Close()
+
+	// Start listening for events.
+	logger.Info("listening for events from watcher")
+	go func() {
+		for {
+			select {
+			case event, ok := <-watcher.Events:
+				if !ok {
+					return
+				}
+				logger.Info("received event", zap.String("event", event.Name))
+				if event.Has(fsnotify.Create) {
+					logger.Info("file created, triggering release", zap.String("event", event.Name))
+					cnsReleaseErr := w.releaseIP(event.Name)
+					if cnsReleaseErr != nil {
+						logger.Error("failed to release IP from CNS", zap.Error(cnsReleaseErr))
+					}
+					deleteErr := RemoveFile(event.Name, path)
+					if deleteErr != nil {
+						logger.Error("failed to remove file", zap.Error(err))
+					}
+				}
+				if event.Has(fsnotify.Remove) {
+					logger.Info("file deleted", zap.String("event", event.Name))
+				}
+			case watcherErr, ok := <-watcher.Errors:
+				if !ok {
+					return
+				}
+				logger.Error("watcher error", zap.Error(watcherErr))
+			}
+		}
+	}()
+
+	// Add directory where intended deletes are kept
+	err = os.MkdirAll(path, 0o755) //nolint
+	if err != nil {
+		logger.Error("error making directory", zap.Error(err))
+	}
```

Comment:
the directory needs to be created here when the watcher starts so that there is a path to watch

Reply:
sure, my concern is that this may hide an issue from us: if the container is not mounting the hostpath correctly, this mkdirall will _create_ the whole path inside the containerfs. we probably don't want that, we want to expect that the path (or most of the path) already exists (because it's been mounted in), and we should fail here instead of creating the whole directory tree if it doesn't exist.

Reply:
so rather than mkdirall, I will update this to mkdir so the only directory that gets created is deleteIDs